### PR TITLE
Feature/tf tools

### PR DESCRIPTION
**Pull request by tajgr**
filter_annotations.py is python script for sorting of images and annotations. Now the output are two csv files (for training and evaluation). In addition, there is a fix of labels/exported_annotations.json file.

### CODE REVIEW

File: annotation-tool/filter_annotations.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            regions = item['regions']
+            for reg in regions:
+                artf_name = reg['region_attributes']['artifact']
+                if artf_name not in USED_ARTF_NAME:
+                    continue
+                x = reg['shape_attributes']['x']
+                y = reg['shape_attributes']['y']
+                width = reg['shape_attributes']['width']
+                height = reg['shape_attributes']['height']
+                data.append( [file_name, artf_name, [x, y, x + width, y+ height], "None"] )
```
Comment (jisa) [inline annotation-tool/filter_annotations.py:24]: What is the extra "None" item?

File: annotation-tool/filter_annotations.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        file_name, artf_name, bbox, use_for = data[ii]
+        x, y, xw, yh = bbox
+        img = cv2.imread(file_name, 1)
+        cv2.rectangle(img, (x,y), (xw, yh), (0,0,255))
+        cv2.putText(img, artf_name, (x,y),fontFace=cv2.FONT_HERSHEY_SIMPLEX, fontScale=1, color=(0,0,255))
+        cv2.putText(img, use_for, (10, 50), fontFace=cv2.FONT_HERSHEY_SIMPLEX, fontScale=1, color=(255, 0, 0))
+        cv2.imshow("win", img)
+
+        k = cv2.waitKey(1) & 0xFF
+        if k == ord("n"): # next img
```
Comment (jisa) [inline annotation-tool/filter_annotations.py:40]: Is one second delay suffiient? What if you get distracted?

File: annotation-tool/filter_annotations.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            ii -= 1
+        elif k == ord("t"):
+            data[ii][3] = "train"
+        elif k == ord("e"):
+            data[ii][3] = "eval"
+        elif k == ord("q"):  # close
+            cv2.destroyAllWindows()
+            break
+
+    cv2.destroyAllWindows()
```
Comment (jisa) [inline annotation-tool/filter_annotations.py:52]: You can only get here after pressing Q, which means that destroyAllWindows was already called. No need to call it twice. Tbh, I would get rid of the first one and keep the one here.

File: annotation-tool/filter_annotations.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                artf_name = reg['region_attributes']['artifact']
+                if artf_name not in USED_ARTF_NAME:
+                    continue
+                x = reg['shape_attributes']['x']
+                y = reg['shape_attributes']['y']
+                width = reg['shape_attributes']['width']
+                height = reg['shape_attributes']['height']
+                data.append( [file_name, artf_name, [x, y, x + width, y+ height], "None"] )
+    ii = 0
+    while True:
```
Comment (jisa) [inline annotation-tool/filter_annotations.py:26]: How concerned are you about an infinite loop when the input is empty?

File: annotation-tool/filter_annotations.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    out_train = open(out_prefix + "_train.csv", "w")
+    out_eval = open(out_prefix + "_eval.csv", "w")
+    out_train.write("filename,class,xmin,ymin,xmax,ymax\r\n")
+    out_eval.write("filename,class,xmin,ymin,xmax,ymax\r\n")
+    for item in data:
+        file_name, artf_name, bbox, use_for = item
+        x, y, xw, yh = bbox
+        if use_for == "train":
+            out_train.write( "%s,%s,%d,%d,%d,%d\r\n" %(file_name, artf_name, x, y, xw, yh))
```
Comment (jisa) [inline annotation-tool/filter_annotations.py:62]: Nit: One space at wrong place and one space probably missing.

You could create the string once before the If statement to avoid copy-pasting the crazy %s%s thing.

Comment (jisa) [inline annotation-tool/filter_annotations.py:62]: Is the \r character needed? I thought it is automatically added in Windows. (And not desired in Linux.)

File: annotation-tool/generate_tfrecord.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+flags.DEFINE_string('label', '', 'Name of class label')
+# if your image has more labels input them as
+# flags.DEFINE_string('label0', '', 'Name of class[0] label')
+# flags.DEFINE_string('label1', '', 'Name of class[1] label')
+# and so on.
+flags.DEFINE_string('img_path', '', 'Path to images')
+FLAGS = flags.FLAGS
+
+
+# TO-DO replace this with label map
```
Comment (jisa) [inline annotation-tool/generate_tfrecord.py:39]: What about doing it straight away? The long IF sequence is unwieldy.

File: labels/exported_annotations.json [hunk 7]
```diff
@@ -9095,8 +9095,8 @@
     ],
     "file_attributes": {}
   },
-  "        ../virtual/helmet/v62A-004.jpg-1": {
```
Comment (jisa) [inline labels/exported_annotations.json:9098]: Conceptually, these changes belong to a separate PR.

File: annotation-tool/filter_annotations.py (hunk 1 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:24]: It is label. This string specifies how the annotation to be used. In the beginning, all is "None" ( do not use).

File: annotation-tool/filter_annotations.py (hunk 4 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:26]: The code going to fall with empty input.

Comment (tajgr) [reply annotation-tool/filter_annotations.py:26]: Maybe I could add something like:
```
    if len(data)==0:
        break
```
but i am not sure if necessary.

File: annotation-tool/filter_annotations.py (hunk 2 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:40]: Actually, it is one millisecond. I get new image only when "n" or "b" key is pressed.

File: annotation-tool/filter_annotations.py (hunk 3 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:52]: Ok, that make sense

File: annotation-tool/filter_annotations.py (hunk 5 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:62]: Ok

Comment (tajgr) [reply annotation-tool/filter_annotations.py:62]: The problem occurs when a file created in Linux (without \r) is opened in Windows. However, I am not sure what happen if somebody run this script in Windows. Blank lines?

File: annotation-tool/generate_tfrecord.py (hunk 6 above)
Comment (tajgr) [reply annotation-tool/generate_tfrecord.py:39]: Yes, this should be generated from "pbtxt" file. I would like to keep it in this form for now and inproved it later. There is still possibility that the tensorflow will be replaced by somethink else.

File: annotation-tool/filter_annotations.py (hunk 1 above)
Comment (jisa) [reply annotation-tool/filter_annotations.py:24]: It may be a good idea to add a comment about that.

File: annotation-tool/filter_annotations.py (hunk 2 above)
Comment (jisa) [reply annotation-tool/filter_annotations.py:40]: Ah, right, OK. In that case ... a longer delay may be more battery-friendly on a laptop :-)

File: annotation-tool/filter_annotations.py (hunk 5 above)
Comment (jisa) [reply annotation-tool/filter_annotations.py]: My best guess is that it should work just fine. Is anybody going to run this on Windows?

File: annotation-tool/filter_annotations.py (hunk 1 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:27]: Done

File: annotation-tool/filter_annotations.py (hunk 2 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py:43]: That is actually really good point.

File: annotation-tool/filter_annotations.py (hunk 5 above)
Comment (tajgr) [reply annotation-tool/filter_annotations.py]: > Is anybody going to run this on Windows?

I dont think so.

**Comment by tajgr** [comment]
Thanks and see last commit.
